### PR TITLE
Https client

### LINKED ISSUE #3: Build Capability-based HTTP client
Create capability-based http client on top of hyper that we can expose to modules.
 
 > capabilities should not require any complicated validity checks to see if some kind of request is valid, and should carry their domain with them implicitly.
capabilities should match reasonable human-relevant security domains
capabilities should minimize the effort required to handle them programmatically
capabilities should be granular/modular/composable as much as possible. 
Thus, the most basic kind of capability is one which has a specific domain, and lets the holder request any HTTPS resource from that domain by the path and query string.
This can be restricted and generalized both up and down.
So, we can have a root level capability that allows sending an https request to any domain with any path; we can initialize this by providing caps to network calls, then use methods on this to get capabilities to specific domains.
Additionally, we can have capabilities that allow only accessing a single path, descendent paths, and/or already have certain query parameters or body values bound. As an example of the use of this, we could hand a capability that grants read only access to the data returned by a particular rest API endpoint, but not the rest of the API or site, with an API key bound so that the thing using the API can't access the key.

> .. is an annoyance in capability security, so we'll be eliminating that.
Rather than specifying full paths, the primary way to provide paths will be an array of components to follow the point which the cap points to. Parsing URLs/URIs into these will be provided as a convenience utility implemented by existing well established libraries.
Similarly, lists of key-value pairs for the query parameters, and blobs for bodies. query parameters can be automatically concatenated, and added at any level of the hierarchy.

```capnp
interface HTTPS {
  domain @0 (name :Text) -> (result :Domain)
}

interface Domain {
  subdomain @0 (name :Text) -> (result :Domain);
  path @1 (name :Text) -> (result :Path);
}

enum HttpVerb {
    GET @0;
    HEAD @1;
    POST @2;
    PUT @3;
    DELETE @4;
    OPTIONS @5;
    PATCH @6;
}

struct KeyValue {
  key @0 :Text;
  value @1 :Text;
}

struct HttpResult {
  statusCode @0 :Int;
  headers @1 :List(KeyValue);
  body @2 :Text;
}



interface Path {
  query @0 (values :List(KeyValue)) -> (result :Path);
  path @1 (values :List(Text)) -> (result :Path);
  get @2 () -> (result :HttpResult);
  head @3 () -> (result :HttpResult);
  post @4 (body :Text) -> (result :HttpResult);
  put @5 (body :Text) -> (result :HttpResult);
  delete @6 (body :Text) -> (result :HttpResult);
  options @7 () -> (result :HttpResult);
  patch @8 (body :Text) -> (result :HttpResult);
  finalize_query @9 -> (result :Path);
  finalize_path @10 -> (result :Path);
  whitelist_verbs @11 (verbs :List(HttpVerb)) -> (result :Path);
  headers @12 (headers: List(KeyValue)) -> (result :Path);
  finalize_headers @13 -> (result :Path);
}
```

`finalize_query` simply returns a copy of the same path with a marker that will throw an error if query() is ever called on it.
`finalize_path` simply returns a copy of the same path wit ha marker that will throw an error if path() is ever called on it.
`whitelist_verbs` return s a copy of the same path where, if `request` is called with a verb that isn't in the whitelist, an error is thrown. By default, all verbs are allowed.
`headers` appends a list of header key value pairs to a path that will be sent along when `request` is eventually called, including on any derived capabilities.
`finalize_headers` returns a copy of the same path with a marker that throws an error if headers() is ever called on it.

### PR DESCRIPTION
**Pull request by ugandalf**
Resolves #3. Adds hyper-tls as a dependency for https connector support. In its current form exposes https client via `https_client()` function. Some tests included in `path.rs`. Schema itself was modified to fix compile errors and structs moved inside `Path` capability as they were only used there.

### CODE REVIEW

File: core/schema/http.capnp [hunk 1]
```diff
@@ -0,0 +1,48 @@
+@0x855e374bb4054d72;
+
+interface Https {
+  domain @0 (name :Text) -> (result :Domain);
+}
+
+interface Domain {
+  subdomain @0 (name :Text) -> (result :Domain);
+  path @1 (value :Text) -> (result :Path);
```
Comment (ErikMcClure) [inline core/schema/http.capnp:9]: Please make this a `:List(Text)` and rename it "values" to reach parity with the path interface

File: core/tests/http_test.rs [hunk 2]
```diff
@@ -0,0 +1,368 @@
...
+    let client = https_client();
+
+    // Requesting domain
+    let mut request = client.domain_request();
+    {
+        request.get().set_name("httpbin.org");
+    }
+    let _ = request.send().promise.await?.get()?.get_result()?;
+
+    ::core::result::Result::Err(::capnp::Error::unimplemented(
```
Comment (ErikMcClure) [inline core/tests/http_test.rs:160]: Please add a "finalize_domain" function to the domain interface that functions like the other finalize functions, and fill out the rest of this test accordingly.

**Review by ErikMcClure — CHANGES_REQUESTED** [review]

File: core/schema/http.capnp (hunk 1 above)
Comment (ugandalf) [reply core/schema/http.capnp]: Did that in 4f05c1e934bad7a07fbd9a3851f7884c8d2e8844 and I consider it resolved.

File: core/tests/http_test.rs (hunk 2 above)
Comment (ugandalf) [reply core/tests/http_test.rs]: "finalize_domain" was done in d361c45614cf91a95b8e828bd4d7180c7b8f628c, tests added in following three commits, but I'm uncertain about how the edge cases should be handled. As it is, paths ending with .. and such are valid, but don't refer to parent directory. Trying to use http escapes results in path ending with "&period%3B&period%3B". Also unsure if that's how protocol relative url breakout would look like, so review would be appreciated. Otherwise I also consider it resolved.

**Review by aiverson — COMMENTED** [review]
This test doesn't actually test what it says it does. This creates a new empty request to add things to the query string, then tests what query strings are in the newly created empty request, which is empty. It does not send anything to the provider and thus can't actually retrieve any data from it. The desired assurance is that there isn't a way to actually retrieve the already set query parameters from a path, to resist attempts to exfiltrate API tokens when handed an API url.

File: core/tests/http_test.rs [hunk 3]
```diff
@@ -0,0 +1,548 @@
...
+    }
+    let path_client = request.send().promise.await?.get()?.get_result()?;
+    let request = path_client.finalize_query_request();
+    let finalized_client = request.send().promise.await?.get()?.get_result()?;
+
+    // Attempting to get query
+    let mut request = finalized_client.query_request();
+    {
+        let values = request.get().get_values()?;
+        assert_eq!(values.len(), 0);
```
Comment (aiverson) [inline core/tests/http_test.rs:289]: this test also doesn't assure anything useful

File: core/tests/http_test.rs [hunk 4]
```diff
@@ -0,0 +1,548 @@
...
+    let path_client = request.send().promise.await?.get()?.get_result()?;
+
+    // Attempting to get root path
+    let mut request = path_client.path_request();
+    {
+        let mut values_builder = request.get().init_values(1);
+        values_builder.reborrow().set(0, "/");
+    }
+    let err = request.send().promise.await;
+
```
Comment (aiverson) [inline core/tests/http_test.rs:446]: this doesn't actually assert anything

File: core/tests/http_test.rs [hunk 5]
```diff
@@ -0,0 +1,548 @@
...
+        let mut values_builder = request.get().init_values(1);
+        values_builder.reborrow().set(0, "anything");
+    }
+    let path_client = request.send().promise.await?.get()?.get_result()?;
+
+    // Attempting to get parent path
+    let mut request = path_client.path_request();
+    {
+        let mut values_builder = request.get().init_values(1);
+        values_builder.reborrow().set(0, "&period;&period;");
```
Comment (aiverson) [inline core/tests/http_test.rs:474]: this is an xml entity and not an http url escape code.

File: core/src/http/path.rs [hunk 6]
```diff
@@ -0,0 +1,591 @@
...
+        let mut request = path_client.whitelist_verbs_request();
+        {
+            let mut whitelist = request.get().init_verbs(2);
+            whitelist.set(0, HttpVerb::Delete);
+            whitelist.set(1, HttpVerb::Get);
+        }
+        assert_eq!(
+            request.send().promise.await.err().map(|e| e.description),
+            Some(
+                "Can't include GET in verb whitelist, because it's not in the original one"
```
Comment (aiverson) [inline core/src/http/path.rs:466]: this error doesn't necessarily make sense.

File: core/tests/http_test.rs (hunk 5 above)
Comment (aiverson) [reply core/tests/http_test.rs:474]: please use the url library's facilities for handling url encoding to verify the safety.

File: core/tests/http_test.rs [hunk 7]
```diff
@@ -0,0 +1,548 @@
...
+    {
+        let mut values_builder = request.get().init_values(1);
+        values_builder.reborrow().set(0, "get");
+    }
+    let path_client = request.send().promise.await?.get()?.get_result()?;
+
+    // Attempting to get path
+    let mut request = path_client.path_request();
+    {
+        let values = request.get().get_values()?;
```
[the remark this answers is not in the document]
Comment (ugandalf) [reply core/tests/http_test.rs:373]: I don't know how to prove impossibility of retrieving previous components via a test, when we don't even have a request that retrieves existing path data. This test was written by Erik and since I don't know what needs to be tested, I left it as-is. Do you want me to remove that test altogether or do you have something in mind?

File: core/tests/http_test.rs (hunk 3 above)
Comment (ugandalf) [reply core/tests/http_test.rs:289]: Is the intention of that test to re-send existing request with changed data? Otherwise analogous to https://github.com/Fundament-Software/keystone/pull/10#discussion_r1134877641

File: core/tests/http_test.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (aiverson) [reply core/tests/http_test.rs:373]: Yes, remove that part of the testing since there isn't a good place to actually test it. Instead, add comments to the capnp schema right next to where any methods that could exfiltrate data would be about how methods which permit the exfiltration need to not be added. That seems like about the best assurance we can do on that part for now.

File: core/tests/http_test.rs (hunk 3 above)
Comment (aiverson) [reply core/tests/http_test.rs:289]: So, what this test is doing is creating an empty request and then checking that the request is empty. The closest thing to that which makes any sense would be trying to retrieve the tests from the remote and verifying that it doesn't provide any. But that's not doable because there is deliberately no method to do so. So it should be treated like the other requirement of not being able to acquire something, and turned into a comment to make a human stop and reconsider since it isn't practical to automatically test it. Perhaps the test can be changed to verify some other piece of functionality about the query strings, or just removed entirely.

File: core/tests/http_test.rs (hunk 4 above)
Comment (aiverson) [reply core/tests/http_test.rs:446]: This is presumably supposed to be checking if putting / on the path permits navigating up the tree, but it needs to assert that whatever it gets is an error from the "/" being an invalid component of a path, or that the result of the get is something other than the expected root item and instead something down the requested path.

**Comment by ugandalf** [comment]
I addressed all the points, I think - url is first constructed with `url` library, then it's converted to string and then that is straight up fed to `hyper::Uri`. In the current state I'm not using `percent-encoding` library - **do you want me to?** I'm also very unsure about corectness / what the intended behavior should be when adding those manually - when manually printing (via `println!("Path: {}", url);` at line 79 of path.rs), I currently get for example paths like `https://httpbin.org/get/%252E%252E?` - `?` added when unnecessary, `%25` being a code for `%`.

Also before merging I think duplication of `lib.rs` and `main.rs` should be resolved, but `lib.rs` can't be simply removed.

**Review by ErikMcClure — APPROVED** [review]
